cmd/customd/client/client.go

```diff
@@ -0,0 +1,448 @@
+package client
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/app"
+	"github.com/iov-one/weave/x/sigs"
+	"github.com/pkg/errors"
+	cmn "github.com/tendermint/tendermint/libs/common"
+	tmpubsub "github.com/tendermint/tendermint/libs/pubsub"
+	"github.com/tendermint/tendermint/rpc/client"
+	ctypes "github.com/tendermint/tendermint/rpc/core/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+)
+
+type Header = tmtypes.Header
+type Status = ctypes.ResultStatus
+type GenesisDoc = tmtypes.GenesisDoc
+
+const BroadcastTxSyncDefaultTimeOut = 15 * time.Second
+
+var QueryNewBlockHeader = tmtypes.EventQueryNewBlockHeader
+
+// Client is an interface to interact with bcp
+type Client interface {
+	TendermintClient() client.Client
+	GetUser(addr weave.Address) (*UserResponse, error)
+	GetWallet(addr weave.Address) (*WalletResponse, error)
+	BroadcastTx(tx weave.Tx) BroadcastTxResponse
+	BroadcastTxAsync(tx weave.Tx, out chan<- BroadcastTxResponse)
+	BroadcastTxSync(tx weave.Tx, timeout time.Duration) BroadcastTxResponse
+	AbciQuery(path string, data []byte) (AbciResponse, error)
+}
+
+// CustomClient is a tendermint client wrapped to provide
+// simple access to the data structures used in bns.
+type CustomClient struct {
+	conn client.Client
+	// subscriber is a unique identifier for subscriptions
+	subscriber string
+}
+
+// NewClient wraps a CustomClient around an existing
+// tendermint client connection.
+func NewClient(conn client.Client) *CustomClient {
+	return &CustomClient{
+		conn: conn,
+		// TODO: make this random
+		subscriber: "tools-client",
+	}
+}
+
+func (cc *CustomClient) TendermintClient() client.Client {
+	return cc.conn
+}
+
+// Nonce has a client/address pair, queries for the nonce
+// and caches recent nonce locally to quickly sign
+type Nonce struct {
```

Comment:
In my opinion the whole nonce can be replaced with a single function to fetch the value. The user can increment this value in the code, without a need for any abstraction for this.

Reply:
Maybe having a struct to wrap all Nonce related methods is not a bad idea. Can you elaborate on why you prefer a single function and the method definition??